tests/admin_views/admin.py

```diff
@@ -1158,3 +1158,31 @@ def has_change_permission(self, request, obj=None):
 
 site10 = admin.AdminSite(name='admin10')
 site10.register(Article, ArticleAdmin10)
+
+
+# Tests #30520
+class ChildForm(forms.ModelForm):
+    extra_field = forms.CharField()
+
+    class Meta:
+        model = Child
+        fields = '__all__'
+
+
+class ChildAdmin(admin.ModelAdmin):
+    '''The ModelForm renders as expected'''
+    form = ChildForm
+
+
+class ChildInline(admin.TabularInline):
+    model = Child
+    form = ChildForm
+
+
+class ParentAdmin(admin.ModelAdmin):
+    inlines = (ChildInline,)
```

Comment:
IMO `ParentAdmin` is not necessary you can always register site with inlines, e.g.
```python
site.register(Parent, inlines=[ChildInlineCustomForm])
```